### PR TITLE
cc_puppet: support AIO installations and more

### PR DESCRIPTION
**Pull request by GabrielNagy**
## Proposed Commit Message

```
cc_puppet: support AIO installations and more

- update the puppet module to support AIO installations by setting
  `install_type` to `aio`
- make the install collection configurable through the `collection`
  parameter; by default the rolling `puppet` collection will be used,
  which installs the latest version)
- when `install_type` is `aio`, puppetlabs repos will be purged after
  installation; set `cleanup` to `False` to prevent this
- AIO installations are performed by downloading and executing a shell
  script; the URL for this script can be overridden using the
  `aio_install_url` parameter
- make it possible to run puppet agent after installation/configuration
  via the `exec` key
- by default, puppet agent will run with the `--test` argument; this can
  be overridden via the `exec_args` key
```

## Additional Context
This was requested on a ticket in the Puppet JIRA: https://tickets.puppetlabs.com/browse/PUP-10926.
The plan is to sign the CLA as an organization, but I opened the PR early to get some review if possible.

## Test Steps
Apart from unit tests, I used an Ubuntu 18.04 machine with QEMU to validate the module functionality; used the following to replace the puppet module and restart the VM:

```sh
virsh shutdown bionic
sleep 1
rm snapshot.img
qemu-img create -f qcow2 -b bionic-server-cloudimg-amd64.img snapshot.img
virt-copy-in -a snapshot.img ~/prepo/cloud-init/cloudinit/config/cc_puppet.py /usr/lib/python3/dist-packages/cloudinit/config
cloud-localds user-data.img user-data
virsh start bionic
```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by GabrielNagy** [comment]
Update: Puppet Labs has signed the Canonical CLA as an organization

**Review by luchihoratiu — APPROVED** [review]

**Review by BobosilaVictor — APPROVED** [review]

**Comment by GabrielNagy** [comment]
Hey @TheRealFalcon, it'd be great to get this reviewed/merged. We validated this internally and everything works as expected, let me know if there's anything else I can do to get the PR merged! Thanks! 😄

File: tests/unittests/test_handler/test_handler_puppet.py [hunk 1]
```diff
@@ -177,3 +267,142 @@ def test_handler_puppet_writes_csr_attributes_file(self, m_subp, m_auto):
               pp_uuid: ED803750-E3C7-44F5-BB08-41A04433FE2E
             """)
         self.assertEqual(expected, content)
+
+    def test_puppet_runs_puppet_if_requested(self, m_auto):
+        """Run puppet with default args if 'exec' is set to True."""
+        with mock.patch.object(
+                subp, 'subp',
```
Comment (TheRealFalcon) [inline tests/unittests/test_handler/test_handler_puppet.py:274]: `autospec` for the mocks in these tests too?

File: tests/unittests/test_handler/test_handler_puppet.py [hunk 2]
```diff
@@ -177,3 +267,142 @@ def test_handler_puppet_writes_csr_attributes_file(self, m_subp, m_auto):
...
+
+        self.assertEqual(
+            [mock.call(blob=response,
+                       args=['-c', 'puppet6-nightly', '--cleanup'],
+                       basename='puppet-install',
+                       capture=False)],
+            m_subp_blob.call_args_list)
+
+    def test_install_with_no_cleanup(self):
+        """Install AIO with specific collection"""
```
Comment (TheRealFalcon) [inline tests/unittests/test_handler/test_handler_puppet.py:393]: `Install AIO with no cleanup`?

File: cloudinit/config/cc_puppet.py [hunk 3]
```diff
@@ -130,23 +214,50 @@ def handle(name, cfg, cloud, log, _args):
...
+    install_type = util.get_cfg_option_str(
+        puppet_cfg, 'install_type', 'packages')
+    cleanup = util.get_cfg_option_bool(puppet_cfg, 'cleanup', True)
+    run = util.get_cfg_option_bool(puppet_cfg, 'exec', default=False)
+    aio_install_url = util.get_cfg_option_str(
+        puppet_cfg, 'aio_install_url', default=AIO_INSTALL_URL)
 
-    p_constants = PuppetConstants(conf_file, ssl_dir, csr_attributes_path, log)
+    # AIO and distro packages use different paths
+    if install_type == 'packages':
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_puppet.py:226]: This clause either needs an else condition or the values defined here need defaults for the case that somebody enters an invalid `install_type`

File: cloudinit/config/cc_puppet.py [hunk 4]
```diff
@@ -119,6 +144,65 @@ def _autostart_puppet(log):
...
+    if version is not None:
+        args = ['-v', version]
+    if collection is not None:
+        args += ['-c', collection]
+
+    # Purge puppetlabs repos after installation
+    if cleanup:
+        args += ['--cleanup']
+    content = url_helper.readurl(url=url, retries=5).contents
+    return subp_blob_in_tempfile(
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_puppet.py:176]: I realize the `subp_blob_in_tempfile` was lifted from the chef module, but the semantics of it are kind of weird. Couldn't we replace this call (and the entire function definition) with something like:
```python
    with temp_utils.tempdir(needs_exe=True) as tmpd:
        tmpf = os.path.join(tmpd, 'puppet-install')
        util.write_file(tmpf, content, mode=0o700)
        return subp.subp([tmpf] + args, capture=False)
```

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Thanks for the submission. Overall this change is good. I left some comments inline, but most are fairly minor.

I noticed that the CLA has been signed on behalf of the company, but additionally any contributor needs to be added (alphabetically) to [the CLA signers file](https://github.com/canonical/cloud-init/blob/master/tools/.github-cla-signers). The full details are described in the last bullet point [of the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#submitting-your-first-pull-request). Currently you only need to add yourself, but feel free to add the other puppet contributors if you have them.

File: tests/unittests/test_handler/test_handler_puppet.py (hunk 1 above)
Comment (GabrielNagy) [reply tests/unittests/test_handler/test_handler_puppet.py]: I changed to mock with decorators, so from what I understand `autospec` shouldn't be needed anymore for these tests, but let me know if I'm wrong.

Comment (TheRealFalcon) [reply tests/unittests/test_handler/test_handler_puppet.py]: Correct

**Review by TheRealFalcon — APPROVED** [review]
Thanks!

**Pull request merged**